rust/osm2lanes/src/transform.rs

```diff
@@ -196,65 +233,157 @@ fn bicycle(
+            }
+        }
+    }
+
+    if tags.is_cycleway(None) {
+        if tags.is_cycleway(Some(WaySide::Both))
+            || tags.is_cycleway(Some(WaySide::Right))
+            || tags.is_cycleway(Some(WaySide::Left))
+        {
+            return Err(LaneSpecError(
+                "cycleway=* not supported with any cycleway:* values".to_owned(),
+            ));
+        }
         forward_side.push(LaneSpec::forward(LaneType::Biking));
-        if !backward_side.is_empty() {
+        if oneway {
+            if !backward_side.is_empty() {
+                // TODO safety check to be checked
+                warnings.0.push(LaneSpecWarning {
+                    description: "oneway has backwards lanes when adding cycleways".to_owned(),
+                    tags: tags.subset(&["oneway", CYCLEWAY]),
+                })
+            }
+        } else {
             backward_side.push(LaneSpec::backward(LaneType::Biking));
         }
-    } else if tags.is_any("cycleway:both", &["lane", "track"]) {
+    } else if tags.is_cycleway(Some(WaySide::Both)) {
         forward_side.push(LaneSpec::forward(LaneType::Biking));
         backward_side.push(LaneSpec::backward(LaneType::Biking));
     } else {
-        // Note here that we look at driving_side frequently, to match up left/right with fwd/back.
-        // If we're driving on the right, then right=fwd. Driving on the left, then right=back.
-        //
-        // TODO Can we express this more simply by referring to a left_side and right_side here?
-        if tags.is_any("cycleway:right", &["lane", "track"]) {
-            if cfg.driving_side == DrivingSide::Right {
-                if tags.is("cycleway:right:oneway", "no") || tags.is("oneway:bicycle", "no") {
-                    forward_side.push(LaneSpec::backward(LaneType::Biking));
-                }
-                forward_side.push(LaneSpec::forward(LaneType::Biking));
-            } else {
-                if tags.is("cycleway:right:oneway", "no") || tags.is("oneway:bicycle", "no") {
-                    backward_side.push(LaneSpec::forward(LaneType::Biking));
-                }
-                backward_side.push(LaneSpec::backward(LaneType::Biking));
-            }
+        // deprecated cycleway=opposite_lane
+        if tags.cycleway_is(None, None, "opposite_lane") {
+            warnings.0.push(LaneSpecWarning {
+                description: "cycleway=opposite_lane deprecated".to_owned(),
+                tags: tags.subset(&[CYCLEWAY]),
+            });
+            backward_side.push(LaneSpec::backward(LaneType::Biking));
         }
-        if tags.is("cycleway:left", "opposite_lane") || tags.is("cycleway", "opposite_lane") {
-            if cfg.driving_side == DrivingSide::Right {
-                backward_side.push(LaneSpec::backward(LaneType::Biking));
-            } else {
-                forward_side.push(LaneSpec::forward(LaneType::Biking));
+        // cycleway:FORWARD=*
```

Comment:
I am still a bit lost (it feels like there are so many combinations to consider), but I think this is becoming more thorough + organized than the original, so thumbs up from me

Reply:
Agreed, it's still not great. I think I want to iterate on the spec before merging something like this. Hopefully then it gets better.